core/iwasm/interpreter/wasm_loader.c

```diff
@@ -250,36 +250,72 @@ loader_malloc(uint64 size, char *error_buf, uint32 error_buf_size)
 static bool
 check_utf8_str(const uint8* str, uint32 len)
 {
-    const uint8 *p = str, *p_end = str + len, *p_end1;
-    uint8 chr, n_bytes;
+    // The valid ranges are taken from page 125, below link
+    // https://www.unicode.org/versions/Unicode9.0.0/ch03.pdf
+    const uint8 *p = str, *p_end = str + len;
+    uint8 chr;
 
     while (p < p_end) {
-        chr = *p++;
-        if (chr >= 0x80) {
-            /* Calculate the byte count: the first byte must be
-               110XXXXX, 1110XXXX, 11110XXX, 111110XX, or 1111110X,
-               the count of leading '1' denotes the total byte count */
-            n_bytes = 0;
-            while ((chr & 0x80) != 0) {
-                chr = (uint8)(chr << 1);
-                n_bytes++;
-            }
-
-            /* Check byte count */
-            if (n_bytes < 2 || n_bytes > 6
-                || p + n_bytes - 1 > p_end)
+        chr = *p;
+        if (chr < 0x80) {
+            p++;
+        }
+        else if (chr >= 0xC2 && chr <= 0xDF) {
+            if (p + 1 >= p_end || p[1] < 0x80 || p[1] > 0xBF) {
```

Comment:
Could we move p + 1 >= p_end to line 263, like other branches?
else if (chr >= 0xC2 && chr <= 0xDF && p + 1 < p_end)